### PR TITLE
UI control group - storage, request, authorization, and unwrapping

### PR DESCRIPTION
**Pull request by meirish**
Note: this is Vault Enterprise Premium feature, though the UI for it is open source.

Now that we can detect control groups in requests in the UI, we've added the ability to look them up (via `sys/control-group/request`), authorize them (via `sys/control-group/authorize`), and unwrap them once authorized.

We customize the flow by comparing the currently logged in token's entityId to the requestor's entity - so we can know if the person looking at it is a requestor. Similarly, we know if the current user has authorized an accessor they're looking up and can change the messaging for them. One example of the customization is if the current user can read the details of the entity in the API, we make the requestor names and the authorization names links to the Identity section of the UI so that they can see more about the users involved.

If you encounter an control group by browsing in the UI, we save off the token, and the URL you were trying to navigate to, and then upon successful authorization, the workflow points them back to that URL in order to view the data in-situ rather than just seeing a JSON blob (a user with a token also has the option of using `Unwrap` in the `Tools` section of the UI if they prefer). If you're using the UI console and get a Control Group token that way, we offer to unwrap the token directly on that page.

Here's what the flow looks like from various perspectives.

Given this control group policy, we'll be looking at a user requesting access that requires 2 authorizations from managers.
screen shot 2018-07-10 at 4 06 15 pm

The accessor lookup form, both requestor and authorizer would see this: 
screen shot 2018-07-10 at 4 06 52 pm

Successful authorizer lookup:
screen shot 2018-07-10 at 4 07 04 pm

Authorizer lookup after authorizing:
screen shot 2018-07-10 at 4 07 10 pm

Requestor lookup (also the screen they are redirected to when encountering a ui control group):
screen shot 2018-07-10 at 4 07 18 pm

Requestor lookup with authorizations:
screen shot 2018-07-10 at 4 07 24 pm

Second authorizer authorization:
screen shot 2018-07-10 at 4 07 54 pm

Requestor fully authorized with a token stored locally:
screen shot 2018-07-10 at 4 09 13 pm

Clicking the "Visit" button before will send them to see the data in the UI:
screen shot 2018-07-10 at 4 08 10 pm

And here's a flow from the console: No redirect just an error in the console with a link:
screen shot 2018-07-10 at 4 10 13 pm

Requestor fully authorized, but no url saved to go back to, so we show them the token form
screen shot 2018-07-10 at 4 11 32 pm

Unwrapping in place:
screen shot 2018-07-10 at 4 11 49 pm

TODO: tests!

### CODE REVIEW

File: ui/app/styles/components/control-group.scss [hunk 1]
```diff
@@ -0,0 +1,48 @@
+.control-group,
+.control-group-success {
+  @extend .box;
```
Comment (johncowen) [inline ui/app/styles/components/control-group.scss]: Ah you've just reminded me of something I have to do, will explain later

File: ui/app/styles/utils/_bulma_variables.scss [hunk 2]
```diff
@@ -17,6 +17,7 @@ $light-blue: #e2eafa;
 $dark-blue: #1563ff;
 $light-blue: rgb(218, 234, 247);
 $dark-green: #36ae40;
+$green-darker: darken($dark-green, 15%);
```
Comment (johncowen) [inline ui/app/styles/utils/_bulma_variables.scss]: Would `$darker-green` be better for consistency?

Little offshoot and just an observation, but this is where `$green-500`, `$green-600` so the numbering thing the design team are doing etc etc would help - means there is plenty of room between 500 and 600 to shim in different shades if we need them post original shades.`$vader-green`? 😄

File: ui/app/templates/components/control-group.hbs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+            </h4>
+            <code class="is-word-break">{{model.id}}</code>
+          </div>
+        </div>
+      {{/if}}
+
+      <div class="authorizations">
+        {{#if (gt model.authorizations.length 0)}}
+        <span class="has-text-success">
+          <ICon @glyph="checkmark-circled-outline" @size={{16}} />
```
Comment (johncowen) [inline ui/app/templates/components/control-group.hbs]: This looks really weird now (`ICon`), at what point can you move this to just `Icon`?

File: ui/app/templates/vault/cluster/access/control-groups.hbs [hunk 4]
```diff
@@ -0,0 +1,32 @@
+<PageHeader as |p|>
+  <p.levelLeft>
+    <h1 class="title is-3">
+      Control Groups
+    </h1>
+  </p.levelLeft>
+  <p.levelRight>
+    {{#link-to 'vault.cluster.settings.control-groups' class="button has-icon-right is-ghost is-compact"}}
+      Configure
+      {{i-con glyph="chevron-right" size=11}}
```
Comment (johncowen) [inline ui/app/templates/vault/cluster/access/control-groups.hbs]: `<ICon>` or `{{i-con}}`? Seems like you are using `<ICon>` above

File: ui/app/templates/components/control-group.hbs (hunk 3 above)
Comment (alisdair) [reply ui/app/templates/components/control-group.hbs]: Good eye! [At least according to the RFC](https://github.com/emberjs/rfcs/blob/master/text/0311-angle-bracket-invocation.md#tag-name), I think we can now use `<Icon>`:

> Another benefit of the capital letter rule is that we can now support component names with a single word, such as <Button>, <Modal> and <Tab>.

Does that work in practice?

Comment (meirish) [reply ui/app/templates/components/control-group.hbs]: Sadly not until the upgrade to 3.2 - single word components is the only omission from the polyfill - https://github.com/rwjblue/ember-angle-bracket-invocation-polyfill#limitations

File: ui/app/templates/vault/cluster/access/control-groups.hbs (hunk 4 above)
Comment (meirish) [reply ui/app/templates/vault/cluster/access/control-groups.hbs]: Whoops! certainly meant to use angle brackets, it's still new so forgot to do it here 😅. I'll get that fixed up.

**Comment by meirish** [comment]
@johncowen yep! more tests are what I'm doing today - there's already acceptance tests for the first PR we did to test intercepting - this will be trickier as it involves a lot of logging in as different users, but hopefully won't be too terrible.

**Comment by meirish** [comment]
Alright, this is ready for a final look - these tests won't run on Travis (we filter out enterprise ones) so here's some screencaps of them running locally:

The control group tests (you can see the filter in the runner):
screen shot 2018-07-12 at 5 15 25 pm

And the full suite: 
screen shot 2018-07-12 at 5 31 47 pm

**Pull request merged**